crypto/x509/x509_vfy.c

```diff
@@ -333,7 +333,7 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
                 || !sk_X509_contains(ctx->chain, issuer))) {
             if (rv == NULL)
                 rv = issuer;
-            if (x509_check_cert_time(ctx, rv, -1))
+            if (x509_check_cert_time(ctx, issuer, -1))
```

Comment:
I suppose you should also return issuer directly here instead of break. As the rv should be returned only in case all check_cert_time calls fail.

Reply:
And one more comment (sorry for that): Let's suppose we have a CA trusted cert store to which we add certificates at the end but never remove the old expired ones. It would be actually better to return the latest cert rather than the first one so I am still not fully convinced this is a right change.

Reply:
> I suppose you should also return issuer directly here instead of break. 

Right - fixed.
I recall that I had already done so, but this must have go lost when I distilled this PR from a larger set of fixes.

Reply:
> Let's suppose we have a CA trusted cert store to which we add certificates at the end but never remove the old expired ones. It would be actually better to return the latest cert rather than the first one so I am still not fully convinced this is a right change.

Well, to judge this properly we need to keep in mind that the preference of first vs. last only applies to certs that are currently invalid (such as, expired). If there is any other matching one that is currently valid, this one will be returned as the best choice.
So otherwise the worst that could happen in the scenario you mentioned is that we return that first expired cert that is older than in the last expired one in the list, so in practice not much of a difference because the verification won't succeed in either case.
If I revert to the original preference of taking the last expired one, which I am open to do if you still believe this is better,
we should document this exception to the normal rule of (in doubt) returning the first cert matching the respective criteria.